File: 1-2-procedures-and-processes-they-generate/1-2-1-linear-recursion-and-iteration/factorial.exs
# SICP 1.2.1
# Linear Recursion and Iteration

defmodule Math do

  def fact1(1), do: 1
  def fact1(n) when n > 1, do: n * fact1(n - 1)

  # Iterative

  def fact2(n) when n >= 1, do: fact2_iter(1, 1, n)
  defp fact2_iter(product, counter, max_count) when counter > max_count, do: product
  defp fact2_iter(product, counter, max_count) do
    fact2_iter((counter * product), (counter + 1), max_count)
  end
  
end

IO.puts(Math.fact1(7))
IO.puts(Math.fact2(7))

File: 1-2-procedures-and-processes-they-generate/1-2-1-linear-recursion-and-iteration/factorial.go
// SICP 1.2.1
// Linear Recursion and Iteration

package main

import "fmt"

func fact1(n int) int {
	if n == 1 {
		return 1
	} else {
		return n * fact1(n-1)
	}
}

// Iterative

func fact2(n int) int {
	return fact2iter(1, 1, n)
}

func fact2iter(product int, counter int, maxcount int) int {
	if counter > maxcount {
		return product
	} else {
		return fact2iter(counter*product, counter+1, maxcount)
	}
}

func main() {
	fmt.Println(fact1(17))
	fmt.Println(fact2(17))
}
